### PR TITLE
Copy retention leases when trim unsafe commits

### PR DESCRIPTION
**Pull request by dnhatn**
When a primary shard is recovered from its store, we trim the last
commit (when it's unsafe). If that primary crashes before the recovery
completes, we will lose the committed retention leases because they are
baked in the last commit. With this change, we copy the retention leases
from the last commit to the safe commit when trimming unsafe commits.

Relates #37165

### CODE REVIEW

**Comment by bleskes** [comment]
This is a getting a bit out of hand. I'm starting to wonder if we're not better off having a dedicated state file in the index/translog/shard folder that is dedicated for this. @dnhatn this is not meant specifically at this PR as you are just extending the "let's store this in a lucene commit" line of thought. I'm just starting to doubt if it is the right one...

**Comment by jasontedor** [comment]
@bleskes Me and @ywelsch had a similar discussion this morning, the complexity is really starting to pile up, that we are basically inventing an entirely new persistence layer with its own semantics.

File: server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java [hunk 1]
```diff
@@ -3815,6 +3819,66 @@ public void testTypelessGet() throws IOException {
         closeShards(shard);
     }
 
+    public void testRecoverFromStoreReserveRetentionLeases() throws Exception {
```
Comment (jasontedor) [inline server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: Can you move this test to IndexShardRetentionLeaseTests?

**Review by jasontedor — APPROVED** [review]
LGTM. I left one request.

**Comment by jasontedor** [comment]
Thanks @dnhatn.

File: server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java (hunk 1 above)
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: I pushed https://github.com/elastic/elasticsearch/pull/37995/commits/f6b270c1d31cc59c211510d28d55eb2f330534f6.

**Comment by dnhatn** [comment]
I am pulling this in, and we will revert it when we have a new persistent layer. Thanks for looking @bleskes and @jasontedor.

**Pull request merged**